File: FateGrandAutomata/DraggableRecyclerView/ItemTouchHelperCallback.cs
using AndroidX.RecyclerView.Widget;

namespace FateGrandAutomata
{
    public class ItemTouchHelperCallback : ItemTouchHelper.Callback
    {
        readonly IItemTouchHelperAdapter _adapter;

        public ItemTouchHelperCallback(IItemTouchHelperAdapter Adapter)
        {
            _adapter = Adapter;
        }

        public override int GetMovementFlags(RecyclerView RecyclerView, RecyclerView.ViewHolder ViewHolder)
        {
            const int dragFlags = ItemTouchHelper.Up | ItemTouchHelper.Down;

            return MakeMovementFlags(dragFlags, 0);
        }

        public override bool IsLongPressDragEnabled => true;

        public override bool IsItemViewSwipeEnabled => false;

        public override bool OnMove(RecyclerView RecyclerView, RecyclerView.ViewHolder ViewHolder, RecyclerView.ViewHolder Target)
        {
            _adapter.OnItemMove(ViewHolder.AdapterPosition, Target.AdapterPosition);
            return true;
        }

        public override void OnSwiped(RecyclerView.ViewHolder ViewHolder, int Direction)
        {
        }

        public override void OnSelectedChanged(RecyclerView.ViewHolder ViewHolder, int ActionState)
        {
            // We only want the active item to change
            if (ActionState != ItemTouchHelper.ActionStateIdle)
            {
                if (ViewHolder is IItemTouchHelperViewHolder itemViewHolder) {
                    // Let the view holder know that this item is being moved or dragged
                    itemViewHolder.OnItemSelected();
                }
            }

            base.OnSelectedChanged(ViewHolder, ActionState);
        }

        public override void ClearView(RecyclerView RecyclerView, RecyclerView.ViewHolder ViewHolder)
        {
            base.ClearView(RecyclerView, ViewHolder);

            if (ViewHolder is IItemTouchHelperViewHolder itemViewHolder) {
                // Tell the view holder it's time to restore the idle state
                itemViewHolder.OnItemClear();
            }
        }
    }
}

File: app/src/main/java/io/github/fate_grand_automata/util/ItemTouchHelperCallback.kt
package io.github.fate_grand_automata.util

import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView

class ItemTouchHelperCallback(
    private val adapter: IItemTouchHelperAdapter,
    private val dragFlags: Int = ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT,
) : ItemTouchHelper.Callback() {

    override fun getMovementFlags(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder): Int {
        return makeMovementFlags(dragFlags, 0)
    }

    override fun isLongPressDragEnabled() = true

    override fun isItemViewSwipeEnabled() = false

    override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
        adapter.onItemMove(
            viewHolder.bindingAdapterPosition,
            target.bindingAdapterPosition,
            viewHolder.itemView,
            target.itemView
        )
        return true
    }

    override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) { }

    override fun onSelectedChanged(viewHolder: RecyclerView.ViewHolder?, actionState: Int) {
        // We only want the active item to change
        if (actionState != ItemTouchHelper.ACTION_STATE_IDLE) {
            if (viewHolder is IItemTouchHelperViewHolder) {
                // Let the view holder know that this item is being moved or dragged
                viewHolder.onItemSelected()
            }
        }

        super.onSelectedChanged(viewHolder, actionState)
    }

    override fun clearView(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder) {
        super.clearView(recyclerView, viewHolder)

        if (viewHolder is IItemTouchHelperViewHolder) {
            // Tell the view holder it's time to restore the idle state
            viewHolder.onItemClear()
        }
    }
}